### ISSUE TITLE
API Reference

### ISSUE THREAD

**Issue by loretoparisi** [spolu/breach_core#114]
Please document then breach/core.

**Comment by spolu** [comment]
Which kind of documentation would you expect to see?

**Comment by loretoparisi** [comment]
All the exposed callback methods like here:

```
  breach.expose('init', function(src, args, cb_) {
  breach.module('core').call('tabs_new_tab_url', { 
    url: 'http://127.0.0.1:' + port + '/newtab'
  }, function(err) {
    console.log('New tab page set! [' + err + ']');
  });
  return cb_();
});
```

**Comment by spolu** [comment]
Yes we urgently need an API reference

**Comment by brayniverse** [comment]
With a bit of guidance I would be happy to help write some documentation. Need a good excuse to familiarise myself with the codebase. @spolu let me know if there's anything I can do.

**Comment by spolu** [comment]
@brayniverse yes that would be awesome. You can start an API reference wiki page on the `breach_core` wiki and we can walk through the code together over IRC.

Basically you want to start out of https://github.com/breach/breach_core/blob/master/lib/core_module.js#L486 that's where the API is exposed to modules.

**Comment by brayniverse** [comment]
@spolu Brilliant. I will have a look at the link you sent me and if possible I will login to IRC after work.

**Comment by spolu** [comment]
@brayniverse that's awesome! Would be really really helpful to have such an API reference!

**Comment by brayniverse** [comment]
@spolu I think a modular, hackable browser is the bees knees, only thing letting it down is how difficult it is getting my head around doing so. I can't wait to start building things for it and writing some docs will definitely help me do so.

**Comment by spolu** [comment]
:+1:

**Comment by brayniverse** [comment]
I'm on IRC at the moment, are you spolu_away?

**Comment by spolu** [comment]
yep

**Comment by loretoparisi** [comment]
+1

**Comment by ghost** [comment]
+1

It'll be great if there's a guidance :)

**Comment by brayniverse** [comment]
@spolu Sorry to spam this thread but it is to do with the API docs.

I can't figure out how to get tab IDs. Most of the `core_tabs` methods require tab IDs and as far as I am aware there is no way of listing all the tab IDs. `tabs_get`lists each tab but the ID is a single digit integer where the methods require something similar to this `881d3dc8c12b8c8a08db935a292d448ceed6c3fc`.

Can you or anybody else provide any guidance?

**Comment by spolu** [comment]
@brayniverse No worry that's fine!

Indeed, in `core_tabs.js`, when using `tabs_get` with no `id` specified, you don't get the tab `id` in the array of tabs returned and that's a bug. You can totally generate a PR on line 691 to add `id: t.id`.

The thing is that tab information can also be received through events handler on the `tabs:state` event emitted by `core`. If you look again in `core_tabs.js` you'll see that as soon as something changes concerning the tabs, the `push_state` function is called which will emit the `tabs:state` event containing all the known information about the tabs, including ids.

Make sense?

**Comment by brayniverse** [comment]
@spolu The variable `t` is defined inside an `if` statement and won't be available within the `else`. I did try to add `my.tabs[id].id` and the `id` parameter passed through the foreach loop and they both have the value '`__NEW_TAB_ID__`'. Can you think of any reason the ID would not be correct?

**Comment by brayniverse** [comment]
Nevermind, I just realised that the module debug page (out) has the ID of '`__NEW_TAB_ID__`'. After opening additional tabs I was able to get the ID's as expected. Will send a PR in a minute.

**Comment by spolu** [comment]
Sweet!

**Comment by brayniverse** [comment]
Hey @spolu I have made some progress with the API reference documentation and have come to the point where it may need some reviewing. There are some parameters missing that I didn't know how to describe them. Could you help me explain the following:
- [text](https://github.com/breach/breach_core/wiki/API-Reference#parameters-7)
- [forward](https://github.com/breach/breach_core/wiki/API-Reference#parameters-7)
- [next](https://github.com/breach/breach_core/wiki/API-Reference#parameters-7)
- [action](https://github.com/breach/breach_core/wiki/API-Reference#parameters-8)
- [element_at](https://github.com/breach/breach_core/wiki/API-Reference#parameters-9)
- [procedure](https://github.com/breach/breach_core/wiki/API-Reference#parameters-10)

Also I noticed that there are three other modules exposed 'controls', 'store' and 'session'. Should I continue to document these as well or are they not ready yet?

**Comment by spolu** [comment]
Ouah! this is awesome! 
- `text` is the search string to perform a find in page action within the targeted tab.
- `forward` is a boolean option to specify if we want to look forward or backward (provided by blink)
- `next` is provided by blink as well and specify if we want to jump to the next result
- action is a string param provided by blink as well. It can be `clear`|`keep`|`activate` by default `clear`
- `procedure` is the procedure name of the module registering as a context_menu builder. Basically the name of the exposed procedure that should be called when the context menu needs to be built
- `element_at` is the coordinates of the element we want to inspect:

```
element_at: {
              x: my.context_menu_params.x,
              y: my.context_menu_params.y
            }
```

controls, and store are definitely ready to be documented!

**Comment by spolu** [comment]
This is so awesome! I'll link to it in the wiki home and breach.cc/hack page!

**Comment by knownasilya** [comment]
@spolu @brayniverse should we add a version of breach that this api applies too? So we can know if things are outdated? Since the API is moving pretty fast.

**Comment by spolu** [comment]
Sure this is `<= 0.3`. Great idea to mention it on the document.
`0.4` will be based on exo_browser `v0.7` and will see its API profoundly changed

**Comment by knownasilya** [comment]
@spolu added.

**Comment by brayniverse** [comment]
Ooh what changes can we be looking forward to seeing?

**Comment by spolu** [comment]
As stated in https://github.com/breach/exo_browser/issues/161, the main changes will be:
- `<webview>` instead of fixed ExoBrowser/ExoFrame layout
- an extended API for window management / menu 

Which means the exo_browser will only expect an HTML document to run and breach will have much more flexibility on what it wants to expose in its API.

As an example you'll be able to probably inject divs into that document as a module and create controls that spans over tabs, are animated, trasparent etc...

We'll also be able to animate the tabs / scale / rotate / etc... lots of fun. A cubic compiz like transition between tabs will be possible where it's not at all at the moment.

**Comment by brayniverse** [comment]
Oh my God so excited!

**Comment by spolu** [comment]
Yes this should be fun :)
